apps/builder/app/builder/builder.tsx

```diff
@@ -208,6 +227,72 @@ const NavigatorPanel = ({
   );
 };
 
+const revealAnimation = ({
+  show,
+  backgroundColor,
+}: {
+  show: boolean;
+  backgroundColor: string;
+}): CSS => ({
+  position: "relative",
+  "> ::after": {
+    content: "",
+    position: "absolute",
+    inset: 0,
+    zIndex: 1,
+    transitionDuration: "300ms",
+    pointerEvents: "none",
+    transitionProperty: "opacity",
+    backgroundColor,
+    opacity: show ? 0 : 1,
+  },
+});
+
+const Loading = ({ value }: { value: number }) => {
+  const [progress, setProgress] = useState(value);
+
+  // This is an approximation of a real progress that should come from "value".
+  useEffect(() => {
+    const timerId = setInterval(() => {
+      setProgress((progress) => {
+        if (progress >= 100) {
+          clearInterval(timerId);
```

Comment:
side effects in set? plz fix

Reply:
I don't like this entire effect, how can it be better written?

Reply:
css keyframes, transitions, some react animate library
as imo any animation should rely on time (RAF https://github.com/streamich/react-use/blob/ade8d3905f544305515d010737b4ae604cc51024/src/useRaf.ts#L4) or css

Reply:
Yes, though in this case, its not controlling each keyframe, that's done by the transition, its just setting rough steps every X ms, with a raf would be calling more state updates than this, I have now made it a bit better, but generally seems fine now